### ISSUE TITLE
IllegalArgumentException when initializing Parse

### ISSUE THREAD

**Issue by hej2010** [parse-community/Parse-SDK-Android#1194]
### New Issue Checklist

- [X] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Android/security/policy).
- [X] I am not just asking a [question](https://github.com/parse-community/.github/blob/main/SUPPORT.md).
- [X] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-Android/issues?q=is%3Aissue).
- [X] I can reproduce the issue with the latest version of [Parse Server](https://github.com/parse-community/parse-server/releases) and the [Parse Android SDK](https://github.com/parse-community/Parse-SDK-Android/releases). 

### Issue Description

The app crashes when initializing Parse with the latest Gradle/Android Studio version when using Proguard/R8 with `minifyEnabled true` and `'com.android.tools.build:gradle:8.0.0'`.

### Steps to reproduce

Initialize Parse at app startup with `Parse.initialize(new Parse.Configuration.Builder(...));`

### Actual Outcome

The app crashes.

### Expected Outcome

The app does not crash.

### Environment

Parse Android SDK
- SDK version: `4.2.0`
- Operating system version: `Android 12`

### Logs

```
Caused by java.lang.IllegalArgumentException: Cannot register a type that does not implement the default constructor!
       at com.parse.ParseObjectSubclassingController.registerSubclass(ParseObjectSubclassingController.java:95)
       at com.parse.ParseObject.registerSubclass(ParseObject.java:291)
       at com.parse.ParseObject.registerParseSubclasses(ParseObject.java:1032)
       at com.parse.Parse.initialize(Parse.java:169)
       at com.parse.Parse.initialize(Parse.java:140)
```
I do not have any custom subclasses registered.

### Fixed by
Keeping all classes that are registered in `registerParseSubclasses()` by adding the line `-keep @com.parse.ParseClassName class com.parse.*` to `proguard-rules.pro`.

**Comment by mtrezza** [comment]
In other words, gradle 8 is not supported OOTB by the Parse SDK, is that correct?

**Comment by hej2010** [comment]
> In other words, gradle 8 is not supported OOTB by the Parse SDK, is that correct?

Yes (if you are using minifyEnabled).

**Comment by mtrezza** [comment]
Pre gradle 8 minify would work, right?

**Comment by hej2010** [comment]
Yes, everything worked until I updated to Gradle 8

**Comment by mtrezza** [comment]
Thanks for clarifying, do you have an idea how to fix it? Would you want to open a PR?

**Comment by hej2010** [comment]
I fixed it by adding
`-keep @com.parse.ParseClassName class com.parse.*`
to proguard-rules.pro, then it no longer crashes.

I´m not sure which file in this project to add it to, maybe https://github.com/parse-community/Parse-SDK-Android/blob/master/parse/release-proguard.pro

**Comment by mtrezza** [comment]
@azlekov, any idea how we could fix this?

**Comment by azlekov** [comment]
No clue, @mtrezza. I will check upcoming days. Maybe is related with latest version of AGP and R8.

**Comment by tSquaredd** [comment]
Experiencing this as well. Proposed fix did not resolve it for us. 
Any ideas?

**Comment by riccardoch** [comment]
Same issue after upgrading Gradle to version 8.
The proposed fix worked for me.

Like @hej2010 I do not have any custom subclasses registered and `minifyEnabled true` for release build.

**Comment by SOGAPPS** [comment]
Currently experiencing this. Proposed fix did not resolve it for us.
We've custom subclasses registered. 

@tSquaredd Did you find a fix?

**Comment by SOGAPPS** [comment]
Thanks @tSquaredd 

Using this worked for us - both "keep" & "keepclassmembers". 

`
-keep @com.parse.ParseClassName class com.parse.*
-keepclassmembers public class * extends com.parse.** {
   public <init>(...);
}
`

**Comment by hej2010** [comment]
Gradle 8.0 switches to using R8 in "full mode" by default: https://r8.googlesource.com/r8/+/refs/heads/master/compatibility-faq.md#r8-full-mode
"_R8 performs more aggressive optimizations, meaning additional ProGuard configuration rules may be required._"

It can be disabled by adding `android.enableR8.fullMode=false` to `gradle.properties`

**Comment by mtrezza** [comment]
Does [this comment](https://github.com/parse-community/Parse-SDK-Android/issues/1194#issuecomment-1595756661) already provide the solution?

@SOGAPPS In your comment did you mean one has to replace `ParseClassName` with the custom classes a developer is using in their app?

**Comment by SOGAPPS** [comment]
Thanks @mtrezza  

No, for clarity, we added the two lines of code to our "proguard rules"

Using both "keep" & "keepclassmembers".
 
`
-keep @com.parse.ParseClassName class com.parse.*
`
`
-keepclassmembers public class * extends com.parse.** {
   public <init>(...);
}
`

**Comment by mtrezza** [comment]
@azlekov do you think a PR with these lines could fix this issue?

**Comment by azlekov** [comment]
I will compare and adjust mine in few days.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [4.2.1](https://github.com/parse-community/Parse-SDK-Android/releases/tag/4.2.1)
